### PR TITLE
[SPARK-9005][MLlib]Fix RegressionMetrics computation of explainedVariance

### PR DESCRIPTION
**Pull request by feynmanliang**
Fixes implementation of `explainedVariance` and `r2` to be consistent with their definitions as described in [SPARK-9005](https://issues.apache.org/jira/browse/SPARK-9005).

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala [hunk 1]
```diff
@@ -53,14 +53,21 @@ class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)]) extend
       )
     summary
   }
+  private lazy val SSerr = math.pow(summary.normL2(1), 2)
+  private lazy val SStot = summary.variance(0) * (summary.count - 1)
+  private lazy val SSreg = {
+    val yMean = summary.mean(0)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala]: Isn't this supposed to be the mean of the data (observations) not predictions?

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala [hunk 2]
```diff
@@ -53,14 +53,21 @@ class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)]) extend
...
+    predictionAndObservations.map {
+      case (prediction, _) => math.pow(prediction - yMean, 2)
+    }.reduce(_ + _)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala]: `.sum`?

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala [hunk 3]
```diff
@@ -53,14 +53,21 @@ class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)]) extend
       )
     summary
   }
+  private lazy val SSerr = math.pow(summary.normL2(1), 2)
+  private lazy val SStot = summary.variance(0) * (summary.count - 1)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala]: Likewise isn't this supposed to be the variance of the observations? col 0 is the predictions right? I feel like I may be missing something.

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala [hunk 4]
```diff
@@ -53,14 +53,21 @@ class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)]) extend
...
   /**
-   * Returns the explained variance regression score.
-   * explainedVariance = 1 - variance(y - \hat{y}) / variance(y)
-   * Reference: [[http://en.wikipedia.org/wiki/Explained_variation]]
+   * Returns the variance explained by regression.
+   * @see [[https://en.wikipedia.org/wiki/Fraction_of_variance_unexplained]]
    */
   def explainedVariance: Double = {
-    1 - summary.variance(1) / summary.variance(0)
+    SSreg / summary.count
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala]: Before this was really _fraction_ of variance unexplained, and that's what your new reference says, but then shouldn't this be `SSreg / SStot`?

Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala]: I updated the reference to one which is about explained/unexplained variance in the context of regression and which also provides explicit formulas for calculation. The calculation before this PR doesn't seem to correspond to anything on either reference.

When the regression model is unbiased (e.g. has an intercept term), the sum of squares can be partitioned (SStot = SSreg + SSerr) and the fraction of variance explained (SSreg / SStot) [is R^2](https://en.wikipedia.org/wiki/Coefficient_of_determination#As_explained_variance).

The [same reference](https://en.wikipedia.org/wiki/Coefficient_of_determination#As_explained_variance) defines explained variance as the variance of the model's predictions (SSreg / n), which I think is more appropriate given that this method is called `explainedVariance` not `proportionVarianceExplained` (which is also a bit redundant with `r2`).

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala (hunk 2 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala]: Thanks!

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala (hunk 4 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala]: Agree with all that. The proportion is kind of redundant. This is going beyond fixing the formula, and also 'fixing' it to return something more consistent with its name. It's an experimental class so seems legitimate. +1

**Comment by feynmanliang** [comment]
jenkins retest this please

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala [hunk 5]
```diff
@@ -53,14 +53,21 @@ class RegressionMetrics(predictionAndObservations: RDD[(Double, Double)]) extend
...
+      case (prediction, _) => math.pow(prediction - yMean, 2)
+    }.sum()
+  }
 
   /**
-   * Returns the explained variance regression score.
-   * explainedVariance = 1 - variance(y - \hat{y}) / variance(y)
-   * Reference: [[http://en.wikipedia.org/wiki/Explained_variation]]
+   * Returns the variance explained by regression.
+   * @see [[https://en.wikipedia.org/wiki/Fraction_of_variance_unexplained]]
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/RegressionMetrics.scala]: Could you put the formula here please?

**Comment by jkbradley** [comment]
LGTM pending tests

**Comment by srowen** [comment]
LGTM

**Comment by jkbradley** [comment]
@feynmanliang Could you please modify the title and description so they just say that explainedVariance is being fixed?  (r2 is not changing, right?)

**Comment by jkbradley** [comment]
Merging with master
Thanks!

**Pull request closed without merging**